packages/env/lib/config/post-process-config.js

```diff
@@ -0,0 +1,195 @@
+'use strict';
+/**
+ * Internal dependencies
+ */
+const mergeConfigs = require( './merge-configs' );
+const addOrReplacePort = require( './add-or-replace-port' );
+const { ValidationError } = require( './validate-config' );
+
+/**
+ * @typedef {import('./config').WPServiceConfig} WPServiceConfig
+ */
+
+/**
+ * Performs any additional post-processing on the config object.
+ *
+ * @param {WPServiceConfig} config The config to process.
+ *
+ * @return {WPServiceConfig} The config after post-processing.
+ */
+module.exports = function postProcessConfig( config ) {
+	// Make sure that we're operating on a config object that has
+	// complete environment configs for convenience.
+	config = mergeRootToEnvironments( config );
+
+	config = appendPortToWPConfigs( config );
+
+	validate( config );
+	return config;
+};
+
+/**
+ * Merges the root config and each environment together in order to make sure each environment has
+ * a full config object to work with internally. This makes it easier than having to try and
+ * resolve the full config options every time we want to use them for something.
+ *
+ * @param {WPServiceConfig} config The config to process.
+ *
+ * @return {WPServiceConfig} The config object with the root options merged together with the environment-specific options.
+ */
+function mergeRootToEnvironments( config ) {
+	// Some root-level options need to be merged early because they have a special
+	// cascade behavior that would break the normal merge. After merging we then
+	// delete them to avoid that breakage and add them back before we return.
+	const removedRootOptions = {};
+	if (
+		config.port !== undefined &&
+		config.env.development.port === undefined
+	) {
+		removedRootOptions.port = config.port;
+		config.env.development.port = config.port;
+		delete config.port;
+	}
+	if (
+		config.testsPort !== undefined &&
+		config.env.tests.port === undefined
+	) {
+		removedRootOptions.testsPort = config.testsPort;
+		config.env.tests.port = config.testsPort;
+		delete config.testsPort;
+	}
+
+	// Merge the root config and the environment configs together so that
+	// we can ignore the root config and have full environment configs.
+	for ( const env in config.env ) {
+		config.env[ env ] = mergeConfigs(
+			deepCopyRootOptions( config ),
+			config.env[ env ]
+		);
+	}
+
+	// Set any root-level options we reset back.
+	for ( const option in removedRootOptions ) {
+		config[ option ] = removedRootOptions[ option ];
```

Comment:
I'm curious about the value of maintaining both the root config and the individual env configs. Do we ever need to access the root config directly internally?

Reply:
In cases like this I think it's important for the output structure to match the input structure. You wouldn't expect post-processing on your config object to discard parts of it. Even if we don't really need this right now, I don't really think there's a problem with maintaining it until the consumer decides they don't need it.

> Do we ever need to access the root config directly internally?

No; it's also discarded almost immediately after this in `loadConfig`. Only the `config.env` keys are passed to the rest of the utility. This won't always be the case though. Once we have root properties like `postInstall`, we will want to bubble those up to `loadConfig()` so it can be put into the `WPConfig` object returned there.